### ISSUE TITLE
Midterm Q4

### ISSUE THREAD

**Issue by pbrahmbhatt3** [SENECA-DSA555/dsa555-w19#38]
bool isPalindrome(const char* s){
        stack <char> Name;
        int n = strlen(s);
        for(int i=0; i<n; i++) {
                Name.push(s[i]);
        }
        int len = Name.size();
          for(int count=0; count<len; count++) {
                if(Name.top() != s[count])
                {
                        return false;
                }
        }
        return true;
}

**Comment by pro2811** [comment]
isPalindrome must be Recursive!! 

````C++

bool isPalindrome(const char *const s, int low, int high) {
    if (high <= low)
        return true;
    else if (s[low] != s[high])
        return false;
    else
        return isPalindrome(s, low + 1, high - 1);
}

bool isPalindrome(const char *const s) {
    return isPalindrome(s, 0, strlen(s) - 1);
}
````

**Comment by ArFe** [comment]
I have a similar but different version:

```c++
bool checkPalindrome(const char* s, int size) {
	if (size <= 0)
		return true;
	else if (s[0] != s[size-1])
		return false;
	else
		checkPalindrome(s+1, (size-2));
}

bool isPalindrome(const char* s) {
	return checkPalindrome(s,strlen(s) - 1);
}
```

**Comment by AZS-99** [comment]
```C++
bool isPalindrome (const char* str) {
    if (strlen(str) < 2)
        return true;
    char* str2 = new char[strlen(str) + 1];
    strncpy(str2, str + 1, strlen(str) - 2);
    return isPalindrome(str2) && (str[0] == str[strlen(str) - 1]);
}
```

**Comment by ArFe** [comment]
> ```c++
> strlen
> ```

I like your solution because you didn't change the function signature.
Although, and I might be wrong, I believe strlen is O(n), and every time you call isPalindrome, it also does strlen(). That, I believe, would make your solution O(n log n) ??? 
(because you decrease the size of the string in every recursion. Not entirely sure though, but seems bigger than O(n) and smaller than O(n^2). Let's see if the professor say something).

**Comment by AZS-99** [comment]
It's O(n) 
For every call, the function copies the passed string but 2 characters. 
That makes the total iterations for copying: n + (n - 2) + (n - 4) + (n - 6) + .... 2 + 1 
Assuming strlen costs the same, when you ADD all of that up, even if it gave you 10000 n, still O(n).

**Comment by ArFe** [comment]
If strlen is O(n) then you have your recursive function O(n) running an O(n) function inside it.
It seems to me
n(n(n(n(n(...(n-n+1)...)-8)-6)-4)-2)
for instance for 8
8(8(8(8(8-8+1)-6)-4)-2) = 8(8(8(8(1)-6)-4)-2) = 8(8(8(2)-4)-2) = 8(8(12)-2) = 8*(94) = 752

Remember, it's a recursive function, so you're going inside, not parallel.

**Comment by cathyatseneca** [comment]
@AZS-99 
your solution is O(n^2)... on any one call it is O(n) because strlen(str) is O(n)... all the C string functions work by looking for '\0' char to find the end.. they don't store size of string so strlen(s) is O(n) to length of s.  Copy is also O(n).

Sooo.. lets suppose you have string of length n.
every call is at least n... you have 3 call to strlen, one to strncpy
So approx
let n represent length of string
T(0) = 2
T(1) = 2
To simplify the math, we'll use your formula
n + (n - 2) + (n - 4) + (n - 6) + .... 2 + 1

This is actually the arithmetic series with every other number removed.
So.. lets call this S.  We'll assume n is odd (the equation won't be different by much if its even)
```
S = n+ (n-2) + (n-4) + ...  5+ 3+1
2S = n+ (n-2) + (n-4) + .... + 1  + (n+ (n-2) + (n-4) + .... + 1)
2S = n + (n-2) + (n-4) + ... +   3   + 1
    +1 +   3      + 5  + ... + (n-2) + n
```
soo... if we look at adding up the numbers up by column first... we end up with:

```
2s = (n+1) + (n+1) + ....(n+1)
```

but that i just adding the same thing over and over... which is same as multiplying.  In total there are n/2 terms of (n+1).  Thus:

```
2S = (n+1)*(n/2)
2S = (n^2)/2 + n/2
S = n^2/4 + n/4
S = 1/4 (n^2) + (1/4)n
```
So... n^2 is the run time.

**Comment by AZS-99** [comment]
Yeah, true! I should have noticed that n + (n - 1) + (n - 2) + ... + 1 = n(n+1)/2.
